### ISSUE TITLE
Lock Timeout even with 60s

### ISSUE THREAD

**Issue by akleber** [frerich/clcache#229]
Before starting to work on the lockfile based locking again, I wanted to see how clcache has evolved since 3.0.3. Also I wanted to get some reference timing data for later comparison with another locking strategy and also wanted to use it on our internal main project.
We have a build driven by ninja (generated by cmake). I am testing on a 4 core VM so there should be at max 4 clcache instances running at the same time. I am using the 3.3.0 release. A complete build generates about 6500 cache entries.
During the build I get a Lock Timeout Error with the 10s default timeout value and even with a 60s timeout.
I then activated profiling to see where the time might be spent. But with profiling on I am not getting any timeouts. The profiling result of a completed run is attached. [prof.txt](https://github.com/frerich/clcache/files/486913/prof.txt)
Any hints on how to diagnose this further? I started looking at the clcache code but it seams that much has changed since I last looked at the code.

**Comment by frerich** [comment]
Hm! First of all, thanks for resuming your work on file-based locking. 👍 

Two things came to my mind while reading your ticket and the `prof.txt` file:

1. It would be great if you could add the line

```python
stats.print_callers()
```

to the end of the `showprofilereport.py` script. It appears that this shows the time spent in each function per caller, this might show who/why so many locks are acquired.

2. I see that clcache was called 6848 times and almost always (6435) it tried to do a direct hit but failed because of an existing manifest file being unusable, i.e. it's basically 100% cache misses. In that case, the original compiler will be invoked (which is where all the time is spent for cache misses, 32087.194  seconds of the 33809.583 seconds overall runtime). At that point, clcache should only have one of the 256 manifest sections locked. So I wonder if for some reason many cache accesses go to the same cache section (which should not happen if the hashes have something resembling a normal distribution...) if there was a bug which somehow caused all invocations to go to the same cache section, then there would be a lot of lock contention.

It might be interesting to add some debug output to the `CacheLock` methods (notably `acquire`) to get an idea of how many concurrent invocations are waiting for a lock, but this will generate a lot of output. Not sure how to get a good report on the lock usage...

**Comment by akleber** [comment]
Sorry I forgot to mention that the description above was with an empty cache. So we had 100% cache misses. In the top level of the cache directory there are 256 directories so the distribution seems as expected.
Do I understand you correctly that clcache now uses 256 different locks for this first level? Also for the stats? This would help for the file based locking I think.
I now did a rebuild with a full cache. This time I got an Lock Timeout even with profiling. Here is the result with your mentioned addition: [prof2.txt](https://github.com/frerich/clcache/files/487275/prof2.txt) The build aborted after about 3200 files.

**Comment by frerich** [comment]
> Do I understand you correctly that clcache now uses 256 different locks for this first level? Also for the stats?

Indeed, clcache uses 256 separate locks depending on the first level to improve concurrency - the stats are still global though and thus guarded by a single lock. In principle this could be changed, but the stats are not locked for very long - and in your case, it doesn't seem to be an issue either.

What I find interesting is that a huge portion of the total time (2439 seconds) is spent in

* 'method 'read' of '_io.BufferedReader' objects' (952 seconds)
* 'method 'acquire' of '_thread.lock' objects' (593 seconds)
* 'built-in method io.open' (156 seconds)
* 'method 'read' of '_io.TextIOWrapper' objects' (144 seconds)

I.e. about 75% of the time is spent in opening files as well as reading them (I suppose this is natural if you have a lot of cache hits, though this could maybe be improved) as well as acquiring thread locks. That locking caught my attention.

To my surprise, almost the entire time (over 99%) spent in `_thread.lock.acqure` is due to calls via `threading.py:1060(_wait_for_tstate_lock)`. `_wait_for_tstate_lock` is called exclusively by `threading.py:1022(join)`, and that function is called exclusively by: `subprocess.py:1283(_communicate)`!

So the entire time waiting for locks is actually caused by using `subprocess.Popen.communicate()`! Looking at the [source code](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L1312) I see that the `subprocess` module forks two threads to read stdout and stderr, and then needs to join them - urgs.

Googling for this phenomenon revealed [various](http://stackoverflow.com/questions/10888846/python-subprocess-module-much-slower-than-commands-deprecated) [pages](http://stackoverflow.com/questions/13835055/python-subprocess-check-output-much-slower-then-call) where people noticed that getting the output of processes executed via `subprocess` is very slow. Interestingly, [one insight](http://stackoverflow.com/a/13837573/91757) was that writing to a temporary file and then reading that was much faster than using a pipe. Maybe we can try that, too (it should be easy, since there is only one place where `communicate` is called in the clcace code.

**Comment by frerich** [comment]
@akleber Can you maybe try the version in #230 ? In the [source code](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L1074) of `subprocess.Popen.communicate` I found this comment which caught my interest:

```python
        # Optimization: If we are not worried about timeouts, we haven't
        # started communicating, and we have one or zero pipes, using select()
        # or threads is unnecessary.
```

Based on that comment, PR #230 adjusts clcache such that reading the process output is not done via two pipes but rather temporary files. Does that help?

**Comment by akleber** [comment]
@frerich I tried #230 but it did not improve the result. Without profiling and 10s timeout clcache still timed out in our build. With a 60s timeout and profiling I was able to build our project with an empty cache [prof3-cold.txt](https://github.com/frerich/clcache/files/493078/prof3-cold.txt) , and I again got an timeout using a full cache. [prof3-warm.txt](https://github.com/frerich/clcache/files/493080/prof3-warm.txt)
The profiling results however look different now.

**Comment by frerich** [comment]
@akleber Thanks for the new results! It's a pity that they seem to have been generated by different builds (2879 invocations of clcache in `prof2.txt` vs. 2073 invocations in `prof3-warm.txt`), so it's a bit hard to compare the absolute performance.

I wonder: when you hit the timeout, do you get a message along the lines of

    Failed to acquire lock <...> after 60000ms; try setting CLCACHE_OBJECT_CACHE_TIMEOUT [..]

Or is it rather

    Error! WaitForSingleObject returns <something>

...or something different altogether?

All this time I was assuming that the timeout being reported is the one we cause by not being able to acquire some cache sections within some time frame -- but that doesn't seem to be the case if I'm reading the profile report correctly, I don't see our `acquire` showing up anywhere!

My first idea was that maybe the `subprocess` module in Python has a default timeout, but that [does not seem to be the case](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait).

**Comment by akleber** [comment]
Sorry I forgot to mention that it always is this message:
`Failed to acquire lock <...> after 60000ms; try setting CLCACHE_OBJECT_CACHE_TIMEOUT [..]`

The timeout does unfortunately not occur at the same place/progress during compilation. The machine is a VM and so fluctuations are expected.
With the 60s timeout and with profiling on the cold cache build without #230 took 117m, the build with the patch took 123m. So assuming a 5% measurement uncertainty the build took the same amount of time.
BTW: I am using Pytnon 3.5.1 32bit.

**Comment by frerich** [comment]
@akleber Thanks for confirming, alas this doesn't make things any less mysterious. :-)

I'm reading [prof3-warm.txt](https://github.com/frerich/clcache/files/493080/prof3-warm.txt ) and trying to make it fit with the observation that some mutex we created couldn't be acquired in time.

The exception you see is raised in response to `WaitForSingleObject` returning a timeout failure code. The only mentioning in the profiler output is `_winapi.WaitForSingleObject` and it shows up very high in the profile output - but the only caller is reported to be `subprocess.py:1260(wait)`, not our `CacheLock.acquire` method. So either I misread (or misunderstand) the profiler output and `WaitForSingleObject` calls are not counted in all cases (??), or we don't actually wait very long but instead have a mistake with checking the return code of the function... or with passing the timeout to the function.

**Comment by frerich** [comment]
@akleber To get a better understanding of how clcache manages the locks in your use case, I now created a new [lock_debug_output](https://github.com/frerich/clcache/tree/lock_debug_output) branch which makes clcache print a *lot* more debug output related to lock management.

If would be much appreciated if you could reproduce your issue with that branch and collect the build output. Afterwards, filter out any lines which do not contain the term `CacheLock` so that you're hopefully only left with the messages related to the cache locks. Then, please upload that log somewhere.

I hope that this is enough to let me reproduce the sequence of lock acquire/release calls which leads to the timeout.

**Comment by akleber** [comment]
Today I began rebaseing my lockfile code to master. While doing so I notice that clcache acquires the object cache lock while it already holds it. This is possible with mutexes and WaitForSingleObject(). It is however not so easy with lockfiles. I have a workaround and during debugging I added a bunch of output.  I am still having timeout problems and looking at my output it seems that the manifest lock is held during compilation. Is that right? In our project we have some huge generated cpp files which take very long to compile, say 10 minutes. So If the lock is held during this time, it is very well possible that another process times out waiting for the lock even though we have 256 possible sections.

**Comment by frerich** [comment]
> I notice that clcache acquires the object cache lock while it already holds it. This is possible with mutexes and WaitForSingleObject(). It is however not so easy with lockfiles.

I imagine you'd use the [`LockFileEx` function](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365203(v=vs.85).aspx) for this, right? There's a little [portalocker](http://code.activestate.com/recipes/65203/) module from where you could extract the Windows-specific portions.

As for recursive locks using `LockFileEx`: I never tried it myself, but the documentation writes:

>  If the locking process opens the file a second time, it cannot access the specified region through this second handle until it unlocks the region.

...so it's clear that *opening* the file a second time won't work: but maybe locking (i.e. calling `LockFileEx`) does?

> [..] looking at my output it seems that the manifest lock is held during compilation. Is that right? In our project we have some huge generated cpp files which take very long to compile, say 10 minutes. So If the lock is held during this time, it is very well possible that another process times out waiting for the lock even though we have 256 possible sections.

Indeed, the manifest section lock is (in case of cache misses) also held during compilation. I'm not very happy with this approach for the reason you outline, and by now I think the only issue is solves is that two concurrent clcache invocations process the exact same compiler invocation (i.e. same source file, same arguments) and end up compiling the file twice (and then clashing when storing the cache entry).

Keeping the manifest section locked is a fairly high price to pay for this exceptional scenario, but I hoped that just having 256 sections makes it so unlikely (1 in 256 chance of collisions) that we won't need to worry about this case. I readily admit that I didn't consider a single compiler invocation taking more than 10 minutes though. :-]

**Comment by akleber** [comment]
I am actually using msvcrt.locking which is using _locking not LockFileEx. Maybe I will have a look at it. But...
The 10 minutes "number" might be a little bit high, so I wanted to analyse this aspect in more detail and added some more debugging output. Strangely enough the last test build went trough without timeout in the warm case (with lockfile). So now I am doing some more builds but as they take 2h in the cold case I will not have any results today. By the way the successful warm build took only 1 hour.

**Comment by akleber** [comment]
So I think I now have a better understanding, why I get the timeouts. I analyzed the last runs with my lockfile implementation, but I think this applies equally to the mutex version.
One of the large generated files did not take 10 minutes to compile but 67 seconds after all. I added timestamps and saw that the manifest lock was acquired and then hold for 67s by one process. Another process by bad luck requested the same lock one second later. As the timeout was 10s, it naturally timed out. So the conclusion here is, that the global timeout has to be at least as long as it takes for the longest compiler invocation. And even then a success is not guaranteed.

Also on one occasion I forgot to make the cache large enough and was surprised to see that the cleanup process takes all 512 manifest and object locks. During this acquiring time it waits until it gets all locks and thereby starting to block other processes until it has acquired all locks and finished its cleanup. I quess this might also become a problem for us at some point.

So what does this all mean for our large project with its peculiarities. It is relatively common that at least two processes want to get the same lock although there are 256 possibilities. In our environment we seem to increase the chance by the fact, that we have some, I would guess about 20, large files that take a minute or even more to compile.
So my next step will be to try with a 10min timeout to get trough our complete build and look at how often a process actually has to wait for a lock and how long.

Our build, which is driven by ninja, is aborted as soon as we hit a timeout. This is not what we want because there was no "real" error. The solution to increase the timeout has the drawback that some few processes are block for an unnecessary amount of time. So might it be possible that in case of a timeout we do not exit clcache with an error code but instead fall back to a simple compiler invocation without the whole cache thing?

**Comment by Liyra** [comment]
I get the same timeout problems with the 3.3.0 :)

**Comment by akleber** [comment]
@frerich I have to admit that I just saw that I missed your message with the lock_debug_output branch. Sorry for that. I think I did more or less the same changes to come to the conclusion of my previous post. So to sum it up I think there is no problem in clcache but rather that I need to use appropriate timeout values for our build.
So for me at least this issue is "invalid".
@Liyra Did you try to use a much larger timeout, say 120s or so?

**Comment by Liyra** [comment]
With 120s, I don't have any timeout. (VS2013 on Windows10 native)

**Comment by frerich** [comment]
Oops, didn't mean to close this when merging #230 ; reopening.

**Comment by frerich** [comment]
I wonder whether the locking right now isn't a bit overzealous. In particular, for a cache miss in 'direct mode' in which there is a matching manifest but that manifest does not have a matching entry, `clcache` will currently execute the steps (indentation indicates that some code is executed while being under a lock):

```
1. Lock one of the 256 manifest sections
    2. Lock one of the 256 compiler artifact sections
        3. Invoke the real compiler (this is the slow part)
        4. Acquire the global statistics lock (this part currently forces all concurrent clcache instances to synchronise)
            5. Update the statistics
            6. Add the generated object file and compiler output to the cache
        8. Unlock the global statistics
    9. Unlock the compiler artifact section
    10. Update the manifest by 'bumping' the matching manifest entry.
11. Unlock the manifest section lock
```

This is correct in that

* the locks are acquired in the correct order (in all of clcache, the statistics lock is only acquired after locking some artifact section, and an artifact section is always acquired after a manifest section lock)
* and it ensures that to the outside, the statistics and the cache contents are always in sync.
* it avoids that any compilation is performed twice, i.e. it will try to reuse as much as possible

However, I think that the timeout issues noticed in this ticket are caused by the original compiler being invoked while some manifest section lock *and* some artifact section lock is held. The fact that we update the statistics and the cache while holding the global statistics lock is not pretty either, but that lock is not held for very long so that's probably not triggering any timeouts.

I'm considering a more 'optimisting' locking strategy which avoids executing the compiler (i.e. doing the slow part) while holding any locks at the expense of potentially compiling a single object file with the exact same flags twice (something which probably happens only very very rarely). The idea is that we do something like:

```
1. Lock manifest and/or artifact section
2. Test if appropriate cache data is available; if data is available, release locks and return data.
3. Release locks
4. Invoke real compiler
5. Acquire manifest and/or artifact section locks again
6. Test if another `clcache` invocation added the cache entry in the meanwhile: if not, add it.
7. Release locks
```

If it's really *only* the (neglible) risk of two clcache instances both noticing that some cached data is unavailable and thus both invoking the real compiler only to then notice that only one of the adds the cache entry, then that would probably help concurrency a lot because we hold the locks for a much shorter duration.

**Comment by frerich** [comment]
@akleber @Liyra It would be much appreciated if you could give PR #244 a shot and see whether you still get the timeout issues (or any other issues) with that.

**Comment by frerich** [comment]
Closing this as #244 was just merged which apparently fixed this issue.

**Comment by TiloW** [comment]
We're still experiencing a similar issue (version 3a91f38b).
We are running up to 8 compiler calls in parallel.

```Failed to execute script clcache
  Traceback (most recent call last):
    File "clcache.py", line 1672, in <module>
    File "clcache.py", line 1509, in main
    File "clcache.py", line 1542, in processCompileRequest
    File "clcache.py", line 1622, in processDirect
    File "clcache.py", line 290, in __enter__
    File "clcache.py", line 314, in acquire
  __main__.CacheLockException: Failed to acquire lock Local\C--Users-ogdf-clcache-manifests-45 after 30000ms; try setti
  ng CLCACHE_OBJECT_CACHE_TIMEOUT_MS environment variable to a larger value.
  Failed to execute script clcache```

**Comment by frerich** [comment]
@TiloW That's interesting! Can you create a new ticket? It may be a different issue with the same symptom. When creating a new ticket, please also see whether increasing the timeout to a large value helps If e.g. 120000 does not help either, then there's probably a deadlock of some sort, which suggests a wrong order of locks being acquired. Otherwise, it might be the (unlikely) case that multiple clcache invocations are accessing the same manifest section (the `45` section) such that one of them failed to get the lock within 30 seconds...
